### PR TITLE
added helpers for work with savepoints

### PR DESCRIPTION
**Pull request by avkhozov**

### CODE REVIEW

**Comment by kraih** [comment]
I've never used savepoints, not sure how useful they are, lets see if more folks are interested.

File: lib/Mojo/Pg/Database.pm [hunk 1]
```diff
@@ -231,6 +231,20 @@ L<Mojo::Pg::Transaction/"commit"> has been called before it is destroyed.
...
+    $db->query('insert into wines values (?, ?)', 10, 'Chateau');
+    $tx->savepoint('sp1');
+    eval { $db->query('insert into wines values (?, ?)', 24, 'Chateau') };
+    if ($@) {
+      $tx->rollback_to('sp1');
+      $db->query('update wines set stock = stock + ? where winename = ?',
+        24, 'Chateau');
+    }
+    $tx->commit;
+  };
```
Comment (kraih) [inline lib/Mojo/Pg/Database.pm]: Not sure upsert is such a great example, considering we will just be able to use a real upsert with PostgreSQL 9.5.

File: lib/Mojo/Pg/Transaction.pm [hunk 2]
```diff
@@ -20,6 +20,10 @@ sub new {
   return $self;
 }
 
+sub savepoint { $_[0]->{dbh}->pg_savepoint($_[1]) }
+
+sub rollback_to { $_[0]->{dbh}->pg_rollback_to($_[1]) }
```
Comment (kraih) [inline lib/Mojo/Pg/Transaction.pm]: These two are in the wrong order.

File: t/results.t [hunk 3]
```diff
@@ -62,6 +62,24 @@ like $@, qr/does_not_exist/, 'right error';
...
+# Savepoints
+my $id;
+{
+  my $tx = $db->begin;
+  $id
+    = $db->query("insert into results_test (name) values ('sp1') returning id")
+    ->array->[0];
+  $tx->savepoint('s1');
+  eval { $db->query("insert into results_test values (?, 'sp2')", $id); };
+  if ($@ && $db->dbh->state == 23505) {
```
Comment (kraih) [inline t/results.t]: Is it safe to use a specific code here?

Comment (avkhozov) [inline t/results.t]: I think these codes (http://www.postgresql.org/docs/9.4/static/errcodes-appendix.html) are permanent and you can use them to better understand what kind of error occurred. In this case this is not matter.

File: lib/Mojo/Pg/Database.pm (hunk 1 above)
Comment (avkhozov) [inline lib/Mojo/Pg/Database.pm]: I mostly use the savepoint for constraint violation. "on duplicate key ignore" in 9.5 will solve this problems. But this is not the only one use case savepoint.

File: t/results.t (hunk 3 above)
Comment (avkhozov) [inline t/results.t]: May be we need a more useful ``Mojo::Pg::Exception`` module which will offer a more convenient interface for handling errors than the comparison dbh state with the code from the documentation?

File: lib/Mojo/Pg/Database.pm (hunk 1 above)
Comment (kraih) [inline lib/Mojo/Pg/Database.pm]: What other use cases are there? Wouldn't one of those make for a much more future-proof example?

Comment (avkhozov) [inline lib/Mojo/Pg/Database.pm]: Any error (division by zero, invalid array index) will destory a whole transaction. Savepoints is a fast method to catch such errors. I use savepoint only in this version, batch insert with restrictions. Example: https://gist.github.com/avkhozov/d6807d3e8f9702a58013.

**Comment by kraih** [comment]
I don't know if savepoints are common enough to deserve their own methods, but i have to give this a :-1: because of the upsert example, which will be pointless after PostgreSQL 9.5.

**Comment by avkhozov** [comment]
Ok, but ``on conflict do`` check only one class of errors.
And I don't understand how to solve what should be in the module ``Mojo::Pg``, and what is not. For example ``pid`` and ``ping`` methods of ``Mojo::Pg::Database`` more useful that savepoints? All these methods are very easy to make using ``dbh``.

**Comment by kraih** [comment]
Think `pid` and `ping` were added because they are needed in tests, which is a strong use case.

**Comment by jhthorsen** [comment]
I don't see the benefit, since it seems easy enough to just do ```$dbcpan ->dbh-> pg_savepoint("sp1")``` directly.

**Pull request closed without merging**